src/lib/updating-element.ts

```diff
@@ -0,0 +1,553 @@
+  }
+
+  /**
+   * Returns the property value for the given attribute value.
+   * Called via the `attributeChangedCallback` and uses the property's `type`
+   * or `type.fromAttribute` property option.
+   */
+  private static _propertyValueFromAttribute(name: PropertyKey, value: string) {
+    const info = this._classProperties.get(name);
+    const type = info && info.type;
+    if (type === undefined) {
+      return value;
+    }
+    const fromAttribute = typeof type === 'function' ? type : type.fromAttribute;
+    return fromAttribute ? fromAttribute(value) : value;
+  }
+
+  /**
+   * Returns the attribute value for the given property value. If this
+   * returns undefined, the property will *not* be reflected to an attribute.
+   * If this returns null, the attribute will be removed, otherwise the
+   * attribute will be set to the value.
+   * This uses the property's `reflect` and `type.toAttribute` property options.
+   */
+  private static _propertyValueToAttribute(name: PropertyKey, value: unknown) {
+    const info = this._classProperties.get(name);
+    if (info === undefined || info.reflect === undefined) {
+      return;
+    }
+    const toAttribute = info.type && (info.type as AttributeSerializer).toAttribute || String;
+    return (typeof toAttribute === 'function') ? toAttribute(value) : null;
+  }
+
+  private _validationState: ValidationState = STATE_IS_VALID;
+  private _instanceProperties: PropertyValues|undefined = undefined;
+  private _validatePromise: Promise<unknown>|undefined = undefined;
+  private _validateResolver: (() => void)|undefined = undefined;
+
+  /**
+   * Object with keys for any properties that have changed since the last
+   * update cycle with previous values.
+   */
+  private _changedProperties: PropertyValues = new Map();
+
+  /**
+   * Node or ShadowRoot into which element DOM should be rendered. Defaults
+   * to an open shadowRoot.
+   */
+  protected renderRoot?: Element|DocumentFragment;
+
+  constructor() {
+    super();
+    this.initialize();
+  }
+
+  /**
+   * Performs element initialization. By default this calls `createRenderRoot` to
+   * create the element `renderRoot` node and captures any pre-set values for
+   * registered properties.
+   */
+  protected initialize() {
+    this.renderRoot = this.createRenderRoot();
+    // Apply any properties set on the instance before upgrade time.
+    for (const p of (this.constructor as typeof UpdatingElement)._classProperties.keys()) {
+      if (this.hasOwnProperty(p)) {
+        const value = this[p as keyof this];
+        delete this[p as keyof this];
+        if (!this._instanceProperties) {
+          this._instanceProperties = new Map();
+        }
+        // NOTE: must capture these into a map and reset at when validating
+        // to avoid stomping on a user value set in the constructor. Being
+        // async doesn't help here since the subclass' constructor value should
+        // be overwritten.
```

Comment:
Is this true? Subclasses can't access `this` until after super, so these assignments will always happen before initializers in the subclass.

Reply:
Assignments are made in the first update cycle.